File: swift/source/core/base/HandlePool.swift
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Foundation

public final class HandlePool {
    private class Wrap {
        let handlePool: HandlePool
        var reference: Int = 0
        init(_ handlePool: HandlePool) {
            self.handlePool = handlePool
        }
    }

    private static let spin = Spin()
    private static var pools: [String: Wrap] = [:]

    static func getPool(withPath path: String, defaultConfigs: Configs) -> RecyclableHandlePool {
        spin.lock(); defer { spin.unlock() }
        var index = pools.index(forKey: path)
        if index == nil {
            let handlePool = HandlePool(withPath: path, defaultConfigs: defaultConfigs)
            pools[path] = Wrap(handlePool)
            index = pools.index(forKey: path)
        }
        return getExistingPool(atIndex: index!)
    }

    static func getExistingPool(with tag: Tag) throws -> RecyclableHandlePool {
        spin.lock(); defer { spin.unlock() }
        guard let index = pools.index(where: { (arg) -> Bool in
            return arg.value.handlePool.tag == tag
        }) else {
            throw Error.reportCore(tag: tag,
                                   path: "",
                                   operation: .getPool,
                                   code: .misuse,
                                   message: "Database with tag: \(tag) is not exists.")
        }
        return getExistingPool(atIndex: index)
    }

    static func getExistingPool(withPath path: String) throws -> RecyclableHandlePool {
        spin.lock(); defer { spin.unlock() }
        guard let index = pools.index(forKey: path) else {
            throw Error.reportCore(tag: nil,
                                   path: path,
                                   operation: .getPool,
                                   code: .misuse,
                                   message: "Database at path: \(path) is not exists.")
        }
        return getExistingPool(atIndex: index)
    }

    private static func getExistingPool(atIndex index: Dictionary<String, Wrap>.Index) -> RecyclableHandlePool {
        let node = pools[index]
        let path = node.key
        var wrap = node.value
        wrap.reference += 1
        return Recyclable(wrap.handlePool, onRecycled: {
            spin.lock(); defer { spin.unlock() }
            let wrap = pools[path]!
            wrap.reference -= 1
            if wrap.reference == 0 {
                pools.removeValue(forKey: path)
            }
        })
    }

    typealias HandleWrap = (handle: Handle, configs: Configs)
    private let handles = ConcurrentList<HandleWrap>(withCapacityCap: maxHardwareConcurrency)

    var tag: Tag?
    let path: String
    private let rwlock = RWLock()
    private let aliveHandleCount = Atomic<Int>(0)
    private var configs: Configs

    private init(withPath path: String, defaultConfigs: Configs) {
        self.path = path
        self.configs = defaultConfigs
    }

    var isDrained: Bool {
        return aliveHandleCount == 0
    }

    func fillOne() throws {
        rwlock.lockRead(); defer { rwlock.unlockRead() }
        let handle = try generate()
        if handles.pushBack(handle) {
            aliveHandleCount += 1
        }
    }

    private func invoke(handleWrap: inout HandleWrap) throws {
        let newConfigs = self.configs
        if newConfigs != handleWrap.configs {
            try newConfigs.invoke(handle: handleWrap.handle)
            handleWrap.configs = newConfigs
        }
    }

    static private let maxConcurrency = max(maxHardwareConcurrency, 64)
    static private let maxHardwareConcurrency = ProcessInfo.processInfo.processorCount

    func flowOut() throws -> RecyclableHandle {
        var unlock = true
        rwlock.lockRead(); defer {  if unlock { rwlock.unlockRead() } }
        var handleWrap = handles.popBack()
        if handleWrap == nil {
            guard aliveHandleCount < HandlePool.maxConcurrency else {
                throw Error.reportCore(tag: tag,
                                       path: path,
                                       operation: .flowOut,
                                       code: .exceed,
                                       message: "The concurrency of database exceeds the max concurrency")
            }
            handleWrap = try generate()
            aliveHandleCount += 1
            if aliveHandleCount > HandlePool.maxHardwareConcurrency {
                var warning = "The concurrency of database: \(tag ?? 0) with \(aliveHandleCount)"
                warning.append(" exceeds the concurrency of hardware: \(HandlePool.maxHardwareConcurrency)")
                Error.warning(warning)
            }
        }
        handleWrap!.handle.tag = self.tag
        try invoke(handleWrap: &handleWrap!)
        unlock = false
        return RecyclableHandle(handleWrap!, onRecycled: {
            self.flowBack(handleWrap!)
        })
    }

    private func flowBack(_ handleWrap: HandleWrap) {
        let inserted = handles.pushBack(handleWrap)
        rwlock.unlockRead()
        if !inserted {
            aliveHandleCount -= 1
        }
    }

    private func generate() throws -> HandleWrap {
        let handle = Handle(withPath: path)
        handle.tag = tag
        let configs = self.configs
        try handle.open()
        try configs.invoke(handle: handle)
        return HandleWrap(handle: handle, configs: configs)
    }

    func blockade() {
        rwlock.lockWrite()
    }

    func unblockade() {
        rwlock.unlockWrite()
    }

    var isBlockaded: Bool {
        return rwlock.isWriting
    }

    public typealias OnDrained = () throws -> Void

    func drain(onDrained: OnDrained) rethrows {
        blockade(); defer { unblockade() }
        let size = handles.clear()
        aliveHandleCount -= size
        try onDrained()
    }

    func drain() {
        blockade(); defer { unblockade() }
        let size = handles.clear()
        aliveHandleCount -= size
    }

    func purgeFreeHandles() {
        rwlock.lockRead(); defer { rwlock.unlockRead() }
        let size = handles.clear()
        aliveHandleCount -= size
    }

    public typealias Config = Configs.Callback
    public typealias ConfigOrder = Configs.Order

    func setConfig(named name: String, with callback: @escaping Config, orderBy order: ConfigOrder) {
        configs.setConfig(named: name, with: callback, orderBy: order)
    }

    func setConfig(named name: String, with callback: @escaping Config) {
        configs.setConfig(named: name, with: callback)
    }

    static func purgeFreeHandlesInAllPools() {
        let handlePools: [HandlePool]!
        do {
            spin.lock(); defer { spin.unlock() }
            handlePools = pools.values.reduce(into: []) { $0.append($1.handlePool) }
        }
        handlePools.forEach { $0.purgeFreeHandles() }
    }
}

File: objc/WCDB/handle_pool.cpp
/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <WCDB/handle.hpp>
#include <WCDB/handle_pool.hpp>
#include <thread>
#include <unordered_map>

namespace WCDB {

std::unordered_map<std::string, std::pair<std::shared_ptr<HandlePool>, int>>
    HandlePool::s_pools;
std::mutex HandlePool::s_mutex;
const int HandlePool::s_hardwareConcurrency =
    std::thread::hardware_concurrency();
const int HandlePool::s_maxConcurrency = 64;

RecyclableHandlePool HandlePool::GetPool(const std::string &path,
                                         const Configs &defaultConfigs)
{
    std::shared_ptr<HandlePool> pool = nullptr;
    std::lock_guard<std::mutex> lockGuard(s_mutex);
    auto iter = s_pools.find(path);
    if (iter == s_pools.end()) {
        pool.reset(new HandlePool(path, defaultConfigs));
        s_pools.insert({path, {pool, 1}});
    } else {
        pool = iter->second.first;
        ++iter->second.second;
    }
    //Can I store iter of unordered_map for later use?
    return RecyclableHandlePool(pool, [](std::shared_ptr<HandlePool> &pool) {
        std::lock_guard<std::mutex> lockGuard(s_mutex);
        const auto &iter = s_pools.find(pool->path);
        if (--iter->second.second == 0) {
            s_pools.erase(iter);
        }
    });
}

RecyclableHandlePool HandlePool::GetPool(Tag tag)
{
    std::lock_guard<std::mutex> lockGuard(s_mutex);
    if (tag != InvalidTag) {
        for (auto iter : s_pools) {
            if (iter.second.first->tag == tag) {
                ++iter.second.second;
                return RecyclableHandlePool(
                    iter.second.first, [](std::shared_ptr<HandlePool> &pool) {
                        std::lock_guard<std::mutex> lockGuard(s_mutex);
                        const auto &iter = s_pools.find(pool->path);
                        if (--iter->second.second == 0) {
                            s_pools.erase(iter);
                        }
                    });
            }
        }
    }
    return RecyclableHandlePool(nullptr, nullptr);
}

void HandlePool::PurgeFreeHandlesInAllPool()
{
    std::list<std::shared_ptr<HandlePool>> handlePools;
    {
        std::lock_guard<std::mutex> lockGuard(s_mutex);
        for (const auto &iter : s_pools) {
            handlePools.push_back(iter.second.first);
        }
    }
    for (const auto &handlePool : handlePools) {
        handlePool->purgeFreeHandles();
    }
}

HandlePool::HandlePool(const std::string &thePath, const Configs &configs)
    : path(thePath)
    , tag(InvalidTag)
    , m_configs(configs)
    , m_handles(s_hardwareConcurrency)
    , m_aliveHandleCount(0)
{
}

void HandlePool::blockade()
{
    m_rwlock.lockWrite();
}

void HandlePool::unblockade()
{
    m_rwlock.unlockWrite();
}

bool HandlePool::isBlockaded() const
{
    return m_rwlock.isWriting();
}

void HandlePool::drain(HandlePool::OnDrained onDrained)
{
    m_rwlock.lockWrite();
    int size = (int) m_handles.clear();
    m_aliveHandleCount -= size;
    if (onDrained) {
        onDrained();
    }
    m_rwlock.unlockWrite();
}

void HandlePool::purgeFreeHandles()
{
    m_rwlock.lockRead();
    int size = (int) m_handles.clear();
    m_aliveHandleCount -= size;
    m_rwlock.unlockRead();
}

bool HandlePool::isDrained()
{
    return m_aliveHandleCount == 0;
}

RecyclableHandle HandlePool::flowOut(Error &error)
{
    m_rwlock.lockRead();
    std::shared_ptr<HandleWrap> handleWrap = m_handles.popBack();
    if (handleWrap == nullptr) {
        if (m_aliveHandleCount < s_maxConcurrency) {
            handleWrap = generate(error);
            if (handleWrap) {
                ++m_aliveHandleCount;
                if (m_aliveHandleCount > s_hardwareConcurrency) {
                    WCDB::Error::Warning(
                        ("The concurrency of database:" +
                         std::to_string(tag.load()) + " with " +
                         std::to_string(m_aliveHandleCount) +
                         " exceeds the concurrency of hardware:" +
                         std::to_string(s_hardwareConcurrency))
                            .c_str());
                }
            }
        } else {
            Error::ReportCore(
                tag.load(), path, Error::CoreOperation::FlowOut,
                Error::CoreCode::Exceed,
                "The concurrency of database exceeds the max concurrency",
                &error);
        }
    }
    if (handleWrap) {
        handleWrap->handle->setTag(tag.load());
        if (invoke(handleWrap, error)) {
            return RecyclableHandle(
                handleWrap, [this](std::shared_ptr<HandleWrap> &handleWrap) {
                    flowBack(handleWrap);
                });
        }
    }

    handleWrap = nullptr;
    m_rwlock.unlockRead();
    return RecyclableHandle(nullptr, nullptr);
}

void HandlePool::flowBack(const std::shared_ptr<HandleWrap> &handleWrap)
{
    if (handleWrap) {
        bool inserted = m_handles.pushBack(handleWrap);
        m_rwlock.unlockRead();
        if (!inserted) {
            --m_aliveHandleCount;
        }
    }
}

std::shared_ptr<HandleWrap> HandlePool::generate(Error &error)
{
    std::shared_ptr<Handle> handle(new Handle(path));
    handle->setTag(tag.load());
    Configs defaultConfigs =
        m_configs; //cache config to avoid multi-thread assigning
    if (!handle->open()) {
        error = handle->getError();
        return nullptr;
    }
    if (defaultConfigs.invoke(handle, error)) {
        return std::shared_ptr<HandleWrap>(
            new HandleWrap(handle, defaultConfigs));
    }
    return nullptr;
}

bool HandlePool::fillOne(Error &error)
{
    m_rwlock.lockRead();
    std::shared_ptr<HandleWrap> handleWrap = generate(error);
    bool result = false;
    if (handleWrap) {
        result = true;
        bool inserted = m_handles.pushBack(handleWrap);
        if (inserted) {
            ++m_aliveHandleCount;
        }
    }
    m_rwlock.unlockRead();
    return result;
}

bool HandlePool::invoke(std::shared_ptr<HandleWrap> &handleWrap, Error &error)
{
    Configs newConfigs =
        m_configs; //cache config to avoid multi-thread assigning
    if (newConfigs != handleWrap->configs) {
        if (!newConfigs.invoke(handleWrap->handle, error)) {
            return false;
        }
        handleWrap->configs = newConfigs;
    }
    error.reset();
    return true;
}

void HandlePool::setConfig(const std::string &name,
                           const Config &config,
                           Configs::Order order)
{
    m_configs.setConfig(name, config, order);
}

void HandlePool::setConfig(const std::string &name, const Config &config)
{
    m_configs.setConfig(name, config);
}

} //namespace WCDB
